File: src/main/java/at/searles/lexer/SkipTokenizer.java
package at.searles.lexer;

import at.searles.lexer.utils.IntSet;
import at.searles.regexp.Regexp;

public class SkipTokenizer implements Tokenizer {

    private final IntSet skippedTokenIds;
    private final Tokenizer parent;

    public SkipTokenizer(Tokenizer parent) {
        this.parent = parent;
        this.skippedTokenIds = new IntSet();
    }

    public Lexer lexer() {
        return parent.lexer();
    }

    public void addSkipped(int tokId) {
        skippedTokenIds.add(tokId);
    }

    @Override
    public IntSet currentTokenIds(TokenStream stream) {
        IntSet currentTokenIds = parent.currentTokenIds(stream);

        while(currentTokenIds != null) {
            int index = currentTokenIds.indexOfFirstMatch(skippedTokenIds);

            if(index == -1) {
                // not a hidden symbol.
                break;
            }

            stream.advance(currentTokenIds.getAt(index));

            currentTokenIds = parent.currentTokenIds(stream);
        }

        return currentTokenIds;
    }

    @Override
    public int add(Regexp regexp) {
        return parent.add(regexp);
    }
}

File: src/main/java/at/searles/lexer/SkipTokenizer.kt
package at.searles.lexer

import at.searles.lexer.utils.IntSet
import at.searles.regexp.Regexp

class SkipTokenizer(override val lexer: Lexer) : Tokenizer {

    private val skippedTokenIds: IntSet = IntSet()

    fun addSkipped(regexp: Regexp): Int {
        val tokenId = add(regexp)
        skippedTokenIds.add(tokenId)
        return tokenId
    }

    override fun add(regexp: Regexp): Int {
        return lexer.add(regexp)
    }

    override fun getCurrentTokenIds(stream: TokenStream): IntSet? {
        var currentTokenIds = lexer.getCurrentTokenIds(stream)

        while (currentTokenIds != null) {
            val index = currentTokenIds.indexOfFirstMatch(skippedTokenIds)

            if (index != -1) {
                stream.advance(currentTokenIds[index])
                currentTokenIds = lexer.getCurrentTokenIds(stream)
            } else {
                break
            }
        }

        return currentTokenIds
    }
}